### ISSUE TITLE
deserialize_str with escaped strings is broken

### ISSUE THREAD

**Issue by jordens** [rust-embedded-community/serde-json-core#74]
`serde_json_core` needs to refuse deserializing a borrowed string with escape sequences.

`serde_json_core::from_slice::<&str>(br#""\n""#)` incorrectly gives `Ok(("\\n", 4))`.

`serde_json::from_slice::<&str>(br#""\n""#)` correctly says `Error("invalid type: string \"\\n\", expected a borrowed string", line: 1, column: 4)`.

**Comment by sammhicks** [comment]
@ryan-summers This is resolved by the merging of #83

**Comment by t-moe** [comment]
Is it fixed?

```
 let mut escape_buf = [0; 8];
 assert_eq!(from_slice_escaped::<&str>(br#""a\nb""#, &mut escape_buf), Ok(("a\nb", 6)));
```

fails with 
```
Err(CustomErrorWithMessage("invalid type: string \"a\\nb\", expected a borrowed string"))
```

while this one is ok:
```
#[derive(Deserialize, PartialEq, Debug)]
  struct Test(heapless::String<10>);

  let mut escape_buf = [0; 8];
  assert_eq!(
      from_slice_escaped::<Test>(br#""a\nb""#, &mut escape_buf),
      Ok((Test(heapless::String::try_from("a\nb").unwrap()), 6))
  );
```

**Comment by ryan-summers** [comment]
You can't deserialize into a `&str` type because the destination data type needs to own the memory for the unescaped string from what I understand. I believe you should be able to do:
```rs
 let mut escape_buf = [0; 8];
 assert_eq!(from_slice_escaped::<heapless::String>(br#""a\nb""#, &mut escape_buf), Ok(("a\nb", 6)));
```

**Comment by t-moe** [comment]
Ah yes. Thank you!

**Comment by jordens** [comment]
It's definitely not fixed.

**Comment by ryan-summers** [comment]
Right now I would classify it as more of an intentional design choice. The `from_slice` and `from_string` APIs currently don't do attempt at string de-escaping. If the user wants to de-escape strings, they can use the new `from_*_escaped` APIs. If we want to directly mirror the `serde-json` API semantics, then indeed this doesn't align with what they do

**Comment by jordens** [comment]
That's not the point. It doesn't need to do de-escaping.
It should error out (as described) instead of performing the wrong deserialization.

**Comment by ryan-summers** [comment]
If the user is explicitly opting in to an API that does not claim to do string escaping, I would expect a deserialization of r#""\n""# to result in a deserialized "\n", or am I misunderstanding something?

Are you referring to the removal of the inner quotes? I'm just a bit confused here

**Comment by t-moe** [comment]
I think a good default for `from_slice` would be to refuse to deserialize strings containing escape sequences.
An API which simply ignores them is unsafe and should be explicitly marked as such. (e.g. `from_slice_ignore_escapes`).

I understand that you dont want to change the behavior `from_slice` now though, as this would be a breaking change... Not sure how to proceed here. But I agree with @jordens that the current behavior is harmful for the users IMHO.

**Comment by jordens** [comment]
The only correct deserialization is one that understands escaping. It's part of Json. If it can't, it must error. Not doing so is a bug, likely even a severe security issue.

**Comment by jordens** [comment]
Could someone please reopen this and mark this as a security vulnerability?

**Comment by ryan-summers** [comment]
I don't have permissions for any security-related issues on this repository. CC @eldruin can you help us out here?

**Comment by eldruin** [comment]
Sorry for the delay.
What do you mean by marking as security vulnerability?
We can add a "security-vulnerability" or similarly-named tag to this issue or publish a full blown security advisory in this repo. Here is the template for that:
```
### Impact
_What kind of vulnerability is it? Who is impacted?_

### Patches
_Has the problem been patched? What versions should users upgrade to?_

### Workarounds
_Is there a way for users to fix or remediate the vulnerability without upgrading?_

### References
_Are there any links users can visit to find out more?_
```
Plus the CVSS scoring.

**Comment by fbartho** [comment]
I know I'm necro-posting, but I wanted to share to any future readers that this issue is still valid:

- The reproducer is still broken on current master (commit a435f788), not just on the published 0.6.0 release. 
- Pull request #83 did not fix the reported issue: it added the opt-in `from_slice_escaped` / `from_str_escaped` API with a caller-supplied unescape buffer, and that API does unescape correctly. 

I'd keep this issue, tracking a change to the default `from_slice` / `from_str` path so it either errors or unescapes instead of returning wrong data.
